File: swimlane_diagram_generator/parser.py
from __future__ import annotations

import re
import shlex
from dataclasses import dataclass

from .model import (
    Connection,
    Diagram,
    Lane,
    Node,
    Shape,
    canonical_symbol_id,
    normalize_identifier,
)

_CONNECTION_WITH_PIPE_LABEL = re.compile(
    r"(?P<src>[A-Za-z_][A-Za-z0-9_-]*)\s*-->\s*\|(?P<label>[^|]+)\|\s*(?P<tgt>[A-Za-z_][A-Za-z0-9_-]*)"
)
_CONNECTION_WITH_OPTIONAL_SUFFIX_LABEL = re.compile(
    r"(?P<src>[A-Za-z_][A-Za-z0-9_-]*)\s*-->\s*(?P<tgt>[A-Za-z_][A-Za-z0-9_-]*)(?:\s*:\s*(?P<label>.+))?"
)


class DiagramSyntaxError(ValueError):
    def __init__(
        self, message: str, line_no: int | None = None, line_text: str | None = None
    ) -> None:
        self.message = message
        self.line_no = line_no
        self.line_text = line_text
        super().__init__(self.__str__())

    def __str__(self) -> str:
        if self.line_no is None:
            return self.message
        excerpt = self.line_text.strip() if self.line_text else ""
        if excerpt:
            return f"Line {self.line_no}: {self.message} -> {excerpt}"
        return f"Line {self.line_no}: {self.message}"


@dataclass(slots=True, frozen=True)
class _ParsedNode:
    node_id: str
    lane_ref: str
    shape: Shape
    text: str


def parse_diagram(source: str) -> Diagram:
    title = ""
    lanes: list[Lane] = []
    nodes: list[Node] = []
    connections: list[Connection] = []

    # phases: 0 = lane declarations, 1 = nodes, 2 = connections
    phase = 0

    lane_id_lookup: dict[str, str] = {}
    lane_ref_lookup: dict[str, str] = {}
    node_lookup: dict[str, Node] = {}

    for line_no, raw_line in enumerate(source.splitlines(), start=1):
        line = raw_line.strip()
        if not line:
            continue
        if line.startswith("%%") or line.startswith("#"):
            continue

        lowered = line.lower()
        if lowered in {"swimlanediagram", "swimlane", "swimlane_flow"}:
            continue

        keyword = lowered.split(maxsplit=1)[0]

        if keyword == "title":
            title = _parse_title(line, line_no, raw_line)
            continue

        if keyword == "lane":
            if phase > 0:
                raise DiagramSyntaxError(
                    "Lane declarations must come before node and connection declarations.",
                    line_no,
                    raw_line,
                )
            lane_id, lane_title = _parse_lane(
                line, line_no, raw_line, lane_index=len(lanes)
            )
            if lane_id in lane_id_lookup:
                raise DiagramSyntaxError(
                    f"Duplicate lane id '{lane_id}'.", line_no, raw_line
                )
            lane = Lane(id=lane_id, title=lane_title, index=len(lanes))
            lanes.append(lane)
            lane_id_lookup[lane.id] = lane.id
            lane_ref_lookup[lane.id] = lane.id
            lane_ref_lookup[lane.title.casefold()] = lane.id
            normalized_title = normalize_identifier(lane.title)
            if normalized_title:
                lane_ref_lookup[normalized_title] = lane.id
            continue

        if keyword == "node":
            if not lanes:
                raise DiagramSyntaxError(
                    "Define at least one lane before nodes.", line_no, raw_line
                )
            if phase == 2:
                raise DiagramSyntaxError(
                    "Node declarations must come before connection declarations.",
                    line_no,
                    raw_line,
                )
            phase = 1
            parsed_node = _parse_node(line, line_no, raw_line)
            lane_id = _resolve_lane_ref(parsed_node.lane_ref, lane_ref_lookup)
            if lane_id is None:
                raise DiagramSyntaxError(
                    f"Unknown lane reference '{parsed_node.lane_ref}' for node '{parsed_node.node_id}'.",
                    line_no,
                    raw_line,
                )
            if parsed_node.node_id in node_lookup:
                raise DiagramSyntaxError(
                    f"Duplicate node id '{parsed_node.node_id}'.", line_no, raw_line
                )

            node = Node(
                id=parsed_node.node_id,
                lane_id=lane_id,
                shape=parsed_node.shape,
                text=parsed_node.text,
                order=len(nodes),
            )
            nodes.append(node)
            node_lookup[node.id] = node
            continue

        if keyword == "connect" or "-->" in line:
            if not nodes:
                raise DiagramSyntaxError(
                    "Define nodes before connection declarations.",
                    line_no,
                    raw_line,
                )
            phase = 2
            connection = _parse_connection(line, line_no, raw_line)
            if connection.source not in node_lookup:
                raise DiagramSyntaxError(
                    f"Connection source node '{connection.source}' does not exist.",
                    line_no,
                    raw_line,
                )
            if connection.target not in node_lookup:
                raise DiagramSyntaxError(
                    f"Connection target node '{connection.target}' does not exist.",
                    line_no,
                    raw_line,
                )
            connections.append(connection)
            continue

        raise DiagramSyntaxError(
            "Unknown statement. Use: title, lane, node, connect, or Mermaid-style 'A --> B'.",
            line_no,
            raw_line,
        )

    if not lanes:
        raise DiagramSyntaxError("Diagram must declare at least one lane.")
    if not nodes:
        raise DiagramSyntaxError("Diagram must declare at least one node.")

    return Diagram(title=title, lanes=lanes, nodes=nodes, connections=connections)


def _split_tokens(line: str, line_no: int, raw_line: str) -> list[str]:
    try:
        return shlex.split(line, comments=False, posix=True)
    except ValueError as exc:
        raise DiagramSyntaxError(
            f"Cannot parse line: {exc}", line_no, raw_line
        ) from exc


def _strip_matching_quotes(value: str) -> str:
    text = value.strip()
    if len(text) >= 2 and text[0] == text[-1] and text[0] in {'"', "'"}:
        return text[1:-1].strip()
    return text


def _parse_title(line: str, line_no: int, raw_line: str) -> str:
    text = line[len("title") :].strip()
    if text.startswith(":"):
        text = text[1:].strip()
    text = _strip_matching_quotes(text)
    if not text:
        raise DiagramSyntaxError("Title cannot be empty.", line_no, raw_line)
    return text


def _parse_lane(
    line: str, line_no: int, raw_line: str, lane_index: int
) -> tuple[str, str]:
    tokens = _split_tokens(line, line_no, raw_line)
    if len(tokens) < 2:
        raise DiagramSyntaxError("Lane statement is incomplete.", line_no, raw_line)

    if len(tokens) == 2:
        candidate = tokens[1]
        if _is_symbol(candidate):
            lane_id = canonical_symbol_id(candidate, kind="Lane id")
            lane_title = candidate
        else:
            lane_id = f"lane_{lane_index + 1}"
            lane_title = candidate
        return lane_id, lane_title

    lane_id = canonical_symbol_id(tokens[1], kind="Lane id")
    lane_title = " ".join(tokens[2:]).strip()
    if not lane_title:
        lane_title = tokens[1]
    return lane_id, lane_title


def _parse_node(line: str, line_no: int, raw_line: str) -> _ParsedNode:
    tokens = _split_tokens(line, line_no, raw_line)
    if len(tokens) < 5:
        raise DiagramSyntaxError(
            "Node syntax must be: node <id> in <lane> [shape] <text>.",
            line_no,
            raw_line,
        )
    if tokens[2].lower() != "in":
        raise DiagramSyntaxError(
            "Expected keyword 'in' in node declaration.", line_no, raw_line
        )

    node_id = canonical_symbol_id(tokens[1], kind="Node id")
    lane_ref = tokens[3]
    remainder = tokens[4:]

    shape = Shape.PROCESS
    first_token = remainder[0]
    if _looks_like_shape_token(first_token):
        parsed_shape = Shape.try_parse(first_token)
        if parsed_shape is None:
            raise DiagramSyntaxError(
                f"Unknown shape '{first_token}'. Allowed: process, decision, subprocess, start/end, document, data.",
                line_no,
                raw_line,
            )
        shape = parsed_shape
        remainder = remainder[1:]

    if not remainder:
        raise DiagramSyntaxError("Node text cannot be empty.", line_no, raw_line)

    text = " ".join(remainder).strip()
    if not text:
        raise DiagramSyntaxError("Node text cannot be empty.", line_no, raw_line)

    return _ParsedNode(node_id=node_id, lane_ref=lane_ref, shape=shape, text=text)


def _parse_connection(line: str, line_no: int, raw_line: str) -> Connection:
    connection_line = line.strip()
    if connection_line.lower().startswith("connect "):
        connection_line = connection_line.split(maxsplit=1)[1].strip()

    match = _CONNECTION_WITH_PIPE_LABEL.fullmatch(connection_line)
    if match:
        source = canonical_symbol_id(match.group("src"), kind="Node id")
        target = canonical_symbol_id(match.group("tgt"), kind="Node id")
        label = match.group("label").strip()
        return Connection(source=source, target=target, label=label or None)

    match = _CONNECTION_WITH_OPTIONAL_SUFFIX_LABEL.fullmatch(connection_line)
    if match:
        source = canonical_symbol_id(match.group("src"), kind="Node id")
        target = canonical_symbol_id(match.group("tgt"), kind="Node id")
        label_raw = match.group("label")
        label = label_raw.strip() if label_raw else None
        return Connection(source=source, target=target, label=label or None)

    raise DiagramSyntaxError(
        "Invalid connection syntax. Use 'connect A --> B', 'A -->|label| B', or 'A --> B : label'.",
        line_no,
        raw_line,
    )


def _resolve_lane_ref(lane_ref: str, lane_ref_lookup: dict[str, str]) -> str | None:
    exact_key = lane_ref.casefold()
    if exact_key in lane_ref_lookup:
        return lane_ref_lookup[exact_key]

    normalized = normalize_identifier(lane_ref)
    if normalized and normalized in lane_ref_lookup:
        return lane_ref_lookup[normalized]

    return None


def _looks_like_shape_token(token: str) -> bool:
    stripped = token.strip()
    if stripped.startswith("[") and stripped.endswith("]"):
        return True
    return Shape.try_parse(stripped) is not None


def _is_symbol(value: str) -> bool:
    return bool(re.fullmatch(r"[A-Za-z_][A-Za-z0-9_-]*", value.strip()))

File: src/parser/parser.ts
// Mirrors swimlane_diagram_generator/parser.py
import {
  canonicalSymbolId,
  isSymbol,
  looksLikeShapeToken,
  normalizeIdentifier,
  parseShape,
  type Connection,
  type Diagram,
  type Lane,
  type Node,
  type Shape,
} from "../model/model.ts";

const CONNECTION_WITH_PIPE_LABEL =
  /^(?<src>[A-Za-z_][A-Za-z0-9_-]*)\s*-->\s*\|(?<label>[^|]+)\|\s*(?<tgt>[A-Za-z_][A-Za-z0-9_-]*)$/;

const CONNECTION_WITH_OPTIONAL_SUFFIX_LABEL =
  /^(?<src>[A-Za-z_][A-Za-z0-9_-]*)\s*-->\s*(?<tgt>[A-Za-z_][A-Za-z0-9_-]*)(?:\s*:\s*(?<label>.+))?$/;

export class DiagramSyntaxError extends Error {
  public readonly lineNo: number | null;
  public readonly lineText: string | null;

  constructor(message: string, lineNo: number | null = null, lineText: string | null = null) {
    super(message);
    this.name = "DiagramSyntaxError";
    this.lineNo = lineNo;
    this.lineText = lineText;
  }

  override toString(): string {
    if (this.lineNo === null) return this.message;
    const excerpt = this.lineText?.trim() ?? "";
    if (excerpt) return `Line ${this.lineNo}: ${this.message} -> ${excerpt}`;
    return `Line ${this.lineNo}: ${this.message}`;
  }
}

interface ParsedNode {
  nodeId: string;
  laneRef: string;
  shape: Shape;
  text: string;
}

export function parseDiagram(source: string): Diagram {
  let title = "";
  const lanes: Lane[] = [];
  const nodes: Node[] = [];
  const connections: Connection[] = [];

  // phases: 0 = lane declarations, 1 = nodes, 2 = connections
  let phase = 0;

  const laneIdLookup: Record<string, string> = {};
  const laneRefLookup: Record<string, string> = {};
  const nodeLookup: Record<string, Node> = {};

  const lines = source.split(/\r?\n/);
  for (let i = 0; i < lines.length; i++) {
    const lineNo = i + 1;
    const rawLine = lines[i]!;
    const line = rawLine.trim();
    if (!line) continue;
    if (line.startsWith("%%") || line.startsWith("#")) continue;

    const lowered = line.toLowerCase();
    if (["swimlanediagram", "swimlane", "swimlane_flow"].includes(lowered)) continue;

    const keyword = lowered.split(/\s+/, 1)[0]!;

    if (keyword === "title") {
      title = parseTitle(line, lineNo, rawLine);
      continue;
    }

    if (keyword === "lane") {
      if (phase > 0) {
        throw new DiagramSyntaxError(
          "Lane declarations must come before node and connection declarations.",
          lineNo,
          rawLine,
        );
      }
      const parsed = parseLane(line, lineNo, rawLine, lanes.length);
      if (laneIdLookup[parsed.laneId]) {
        throw new DiagramSyntaxError(
          `Duplicate lane id '${parsed.laneId}'.`,
          lineNo,
          rawLine,
        );
      }
      const lane: Lane = { id: parsed.laneId, title: parsed.laneTitle, index: lanes.length };
      lanes.push(lane);
      laneIdLookup[lane.id] = lane.id;
      laneRefLookup[lane.id] = lane.id;
      laneRefLookup[lane.title.toLowerCase()] = lane.id;
      const normalizedTitle = normalizeIdentifier(lane.title);
      if (normalizedTitle) laneRefLookup[normalizedTitle] = lane.id;
      continue;
    }

    if (keyword === "node") {
      if (lanes.length === 0) {
        throw new DiagramSyntaxError(
          "Define at least one lane before nodes.",
          lineNo,
          rawLine,
        );
      }
      if (phase === 2) {
        throw new DiagramSyntaxError(
          "Node declarations must come before connection declarations.",
          lineNo,
          rawLine,
        );
      }
      phase = 1;
      const parsedNode = parseNode(line, lineNo, rawLine);
      const laneId = resolveLaneRef(parsedNode.laneRef, laneRefLookup);
      if (laneId === null) {
        throw new DiagramSyntaxError(
          `Unknown lane reference '${parsedNode.laneRef}' for node '${parsedNode.nodeId}'.`,
          lineNo,
          rawLine,
        );
      }
      if (nodeLookup[parsedNode.nodeId]) {
        throw new DiagramSyntaxError(
          `Duplicate node id '${parsedNode.nodeId}'.`,
          lineNo,
          rawLine,
        );
      }
      const node: Node = {
        id: parsedNode.nodeId,
        laneId,
        shape: parsedNode.shape,
        text: parsedNode.text,
        order: nodes.length,
      };
      nodes.push(node);
      nodeLookup[node.id] = node;
      continue;
    }

    if (keyword === "connect" || line.includes("-->")) {
      if (nodes.length === 0) {
        throw new DiagramSyntaxError(
          "Define nodes before connection declarations.",
          lineNo,
          rawLine,
        );
      }
      phase = 2;
      const connection = parseConnection(line, lineNo, rawLine);
      if (!nodeLookup[connection.source]) {
        throw new DiagramSyntaxError(
          `Connection source node '${connection.source}' does not exist.`,
          lineNo,
          rawLine,
        );
      }
      if (!nodeLookup[connection.target]) {
        throw new DiagramSyntaxError(
          `Connection target node '${connection.target}' does not exist.`,
          lineNo,
          rawLine,
        );
      }
      connections.push(connection);
      continue;
    }

    throw new DiagramSyntaxError(
      "Unknown statement. Use: title, lane, node, connect, or Mermaid-style 'A --> B'.",
      lineNo,
      rawLine,
    );
  }

  if (lanes.length === 0) {
    throw new DiagramSyntaxError("Diagram must declare at least one lane.");
  }
  if (nodes.length === 0) {
    throw new DiagramSyntaxError("Diagram must declare at least one node.");
  }

  return { title, lanes, nodes, connections };
}

function splitTokens(line: string, lineNo: number, rawLine: string): string[] {
  // Minimal POSIX-like splitter:
  //  - whitespace separates tokens
  //  - "..." and '...' are kept as a single token with quotes stripped
  // The Python implementation uses shlex; this is a deliberately small port that
  // covers the examples and DSL tests.
  const tokens: string[] = [];
  let i = 0;
  while (i < line.length) {
    const ch = line[i]!;
    if (ch === " " || ch === "\t") {
      i++;
      continue;
    }
    if (ch === '"' || ch === "'") {
      const quote = ch;
      let j = i + 1;
      let buf = "";
      while (j < line.length && line[j] !== quote) {
        if (line[j] === "\\" && j + 1 < line.length) {
          buf += line[j + 1]!;
          j += 2;
        } else {
          buf += line[j]!;
          j++;
        }
      }
      if (j >= line.length) {
        throw new DiagramSyntaxError(`Unmatched ${quote} in line.`, lineNo, rawLine);
      }
      tokens.push(buf);
      i = j + 1;
      continue;
    }
    let j = i;
    while (j < line.length && line[j] !== " " && line[j] !== "\t") {
      j++;
    }
    tokens.push(line.slice(i, j));
    i = j;
  }
  return tokens;
}

function stripMatchingQuotes(value: string): string {
  const text = value.trim();
  if (text.length >= 2 && text[0] === text[text.length - 1] && (text[0] === '"' || text[0] === "'")) {
    return text.slice(1, -1).trim();
  }
  return text;
}

function parseTitle(line: string, lineNo: number, rawLine: string): string {
  let text = line.slice("title".length).trim();
  if (text.startsWith(":")) text = text.slice(1).trim();
  text = stripMatchingQuotes(text);
  if (!text) {
    throw new DiagramSyntaxError("Title cannot be empty.", lineNo, rawLine);
  }
  return text;
}

function parseLane(
  line: string,
  lineNo: number,
  rawLine: string,
  laneIndex: number,
): { laneId: string; laneTitle: string } {
  const tokens = splitTokens(line, lineNo, rawLine);
  if (tokens.length < 2) {
    throw new DiagramSyntaxError("Lane statement is incomplete.", lineNo, rawLine);
  }
  if (tokens.length === 2) {
    const candidate = tokens[1]!;
    if (isSymbol(candidate)) {
      return { laneId: canonicalSymbolId(candidate, "Lane id"), laneTitle: candidate };
    }
    return { laneId: `lane_${laneIndex + 1}`, laneTitle: candidate };
  }
  const laneId = canonicalSymbolId(tokens[1]!, "Lane id");
  const laneTitle = tokens.slice(2).join(" ").trim();
  if (!laneTitle) return { laneId, laneTitle: tokens[1]! };
  return { laneId, laneTitle };
}

function parseNode(line: string, lineNo: number, rawLine: string): ParsedNode {
  const tokens = splitTokens(line, lineNo, rawLine);
  if (tokens.length < 5) {
    throw new DiagramSyntaxError(
      "Node syntax must be: node <id> in <lane> [shape] <text>.",
      lineNo,
      rawLine,
    );
  }
  if (tokens[2]!.toLowerCase() !== "in") {
    throw new DiagramSyntaxError(
      "Expected keyword 'in' in node declaration.",
      lineNo,
      rawLine,
    );
  }

  const nodeId = canonicalSymbolId(tokens[1]!, "Node id");
  const laneRef = tokens[3]!;
  let remainder = tokens.slice(4);

  let shape: Shape = "process";
  const firstToken = remainder[0]!;
  if (looksLikeShapeToken(firstToken)) {
    let parsedShape;
    try {
      parsedShape = parseShape(firstToken);
    } catch {
      throw new DiagramSyntaxError(
        `Unknown shape '${firstToken}'. Allowed: process, decision, subprocess, start/end, document, data.`,
        lineNo,
        rawLine,
      );
    }
    shape = parsedShape;
    remainder = remainder.slice(1);
  }

  if (remainder.length === 0) {
    throw new DiagramSyntaxError("Node text cannot be empty.", lineNo, rawLine);
  }

  const text = remainder.join(" ").trim();
  if (!text) {
    throw new DiagramSyntaxError("Node text cannot be empty.", lineNo, rawLine);
  }
  return { nodeId, laneRef, shape, text };
}

function parseConnection(line: string, lineNo: number, rawLine: string): Connection {
  let connectionLine = line.trim();
  if (connectionLine.toLowerCase().startsWith("connect ")) {
    connectionLine = connectionLine.slice("connect ".length).trim();
  }

  const m1 = CONNECTION_WITH_PIPE_LABEL.exec(connectionLine);
  if (m1 && m1.groups) {
    const source = canonicalSymbolId(m1.groups["src"]!, "Node id");
    const target = canonicalSymbolId(m1.groups["tgt"]!, "Node id");
    const label = m1.groups["label"]!.trim();
    return { source, target, label: label || null };
  }

  const m2 = CONNECTION_WITH_OPTIONAL_SUFFIX_LABEL.exec(connectionLine);
  if (m2 && m2.groups) {
    const source = canonicalSymbolId(m2.groups["src"]!, "Node id");
    const target = canonicalSymbolId(m2.groups["tgt"]!, "Node id");
    const labelRaw = m2.groups["label"];
    const label = labelRaw ? labelRaw.trim() : null;
    return { source, target, label: label || null };
  }

  throw new DiagramSyntaxError(
    "Invalid connection syntax. Use 'connect A --> B', 'A -->|label| B', or 'A --> B : label'.",
    lineNo,
    rawLine,
  );
}

function resolveLaneRef(laneRef: string, laneRefLookup: Record<string, string>): string | null {
  const exactKey = laneRef.toLowerCase();
  if (exactKey in laneRefLookup) return laneRefLookup[exactKey]!;

  const normalized = normalizeIdentifier(laneRef);
  if (normalized && normalized in laneRefLookup) return laneRefLookup[normalized]!;
  return null;
}
